controller/lib/hal/stepper.h

```diff
@@ -0,0 +1,342 @@
+  ELE_POS = 0x02,           // Electrical position
+  MARK_POS = 0x03,          // Mark position
+  SPEED = 0x04,             // Current speed
+  ACCEL = 0x05,             // Acceleration
+  DECEL = 0x06,             // Deceleration
+  MAX_SPEED = 0x07,         // Maximum allowed speed
+  MIN_SPEED = 0x08,         // Minimum allowed speed
+  KVAL_HOLD = 0x09,         //
+  KVAL_RUN = 0x0A,          //
+  KVAL_ACCEL = 0x0B,        //
+  KVAL_DECEL = 0x0C,        //
+  INT_SPEED = 0x0D,         //
+  START_SLOPE = 0x0E,       //
+  FINAL_SLOPE_ACCEL = 0x0F, //
+  FINAL_SLOPE_DECEL = 0x10, //
+  THERMAL_COMP = 0x11,      //
+  ADC_OUT = 0x12,           //
+  OVER_CRNT_THRESH = 0x13,  //
+  STALL_THRESH = 0x14,      //
+  FULL_STEP_SPEED = 0x15,   // Speed at which to switch to full step mode
+  STEP_MODE = 0x16,         //
+  ALARM_ENA = 0x17,         //
+  GATE_CFG1 = 0x18,         //
+  GATE_CFG2 = 0x19,         //
+  CONFIG = 0x1A,            //
+  STATUS = 0x1B,            //
+};
+
+// Error codes returned by my functions
+enum class StepMtrErr {
+  OK,
+  BAD_PARAM,     // The parameter ID is invalid
+  BAD_VALUE,     // Illegal value passed
+  WOULD_BLOCK,   // Call would block, can't be called from within ISR
+  QUEUE_FULL,    // Can't add command to queue, not enough space
+  INVALID_STATE, // Invalid state for command.
+};
+
+// States for the stepper motor communication interface
+enum class StepCommState {
+  IDLE,        // Not currently communicating
+  SEND_QUEUED, // Sending data that was queued up by control loop
+  SEND_SYNC,   // Sending data that the background thread is waiting on.
+};
+
+enum class StepMoveStatus {
+  STOPPED,
+  ACCELERATING,
+  DECELERATING,
+  CONSTANT_SPEED,
+};
+
+// Detailed status about the stepper driver chip.
+// Note, fields marked with * below are latched meaning that they
+// will return true if the event has occurred since the last time
+// the status was read.  The act of reading the status clears them
+// so they won't be true on the next read unless they happen again.
+struct StepperStatus {
+
+  // If true, power is being applied to the motor.
+  bool enabled;
+
+  // * True if the stepper chip has detected an under voltage
+  bool under_voltage;
+
+  // * True if the stepper chip is getting hot
+  bool thermal_warning;
+
+  // * True if the stepper chip got so hot is shut itself down.
+  bool thermal_shutdown;
+
+  // * True if over current has been detected
+  bool over_current;
+
+  // * True if the chip detects a loss of steps (motor stall)
+  bool step_loss;
+
+  // * True if a command error (bad command) was detected by the chip
+  bool command_error;
+
+  // Specifics of the move status
+  StepMoveStatus move_status;
+};
+
+// Represents one of the stepper motors in the system
+class StepMotor {
+
+  // This constant represents the number of motors wired in
+  // to the system.  It needs to match the hardware or the
+  // interface wont work.
+  static const int total_motors_ = 1;
```

Comment:
constexpr.  I had to Google that.  I really haven't been keeping up with the C++ language.
Yes, I'll change it.

Reply:
> Also, I might default initialize with {} rather than =

If you mean the difference between `total_motors_ = 1` and `total_motors_{1}`, these are semantically identical for ints.  In fact I believe they're semantically identical for all types, except that sometimes only the one with braces will compile and the one with an equal sign won't.  That is, if they both compile, they should emit identical code.

One can always check at gcc.godbolt.org